### DISCUSSION TITLE
Transitioning from another replication/snapshot tool?

### DISCUSSION THREAD

**Discussion by gdevenyi** [zrepl/zrepl#567] [Q&A] [answered]
I have two servers doing a 1-1 full zpool replication for disaster recovery, currently via sanoid:
```
for filesystem in $(zfs list -s used -o name -H | grep -- '/'); do
./sanoid/syncoid --compress=zstd-fast --identifier=migrate --recursive \
      --pv-options='-p -t -e -r -b -a' --force-delete \
        --mbuffer-size=1G ${filesystem} root@targethost:${filesystem}
done
```

I also have snapshots with retention via the classic zfs-auto-snapshot on the sender.

I'd like to transition to using zrepl for both features, but I'm unsure how zrepl will interact with this existing setup, and I don't see a FAQ about this (I suggest this, as I suspect many users may be coming from other solutions).

Are there any pitfalls to be aware of transitioning to zrepl?

**Comment by problame** [comment, marked as answer]
I'm not familiar with syncoid. Is it correct that, on the `targethost`, there are only snapshots called `@migrate...`?

**Reply by gdevenyi** [reply]
syncoid is part of the sanoid tools, a much earlier set of management stuff, I was only using syncoid, https://github.com/jimsalterjrs/sanoid

>  there are only snapshots called @migrate...?

There are these existing snapshots on both source and receiver (it really is a 1-1 copy right now)
```
@syncoid_migrate
@syncoid_<other mirror server>
@zfs-auto-snap_<etc>
```

**Reply by gdevenyi** [reply]
Following up here, this is strongly related to #569 as I'm doing that, but I'm already "mostly done" and I would like zrepl to take over this.

**Reply by problame** [reply]
Ok, so, if it's basically like #569 , you could follow the the instructions there. Bascially, define a sink job on your receiving side, with `root_fs: `  set to whatever sub-tree you receive the sending side datasets into. You might have to use `zfs rename` to move around receive-side datasets a little. 

Please report whether this solves your use case.

**Reply by gdevenyi** [reply]
> Bascially, define a sink job on your receiving side, with root_fs: set to whatever sub-tree you receive the sending side datasets into. You might have to use zfs rename to move around receive-side datasets a little

If I understand this correctly, I can't directly sync the pool root 1:1 from production to backup? I have not been using a sub-tree for the replication.

**Reply by gdevenyi** [reply]
Poke @problame. Will I need to rename my existing syncs?

**Reply by InsanePrawn** [reply]
> > Bascially, define a sink job on your receiving side, with root_fs: set to whatever sub-tree you receive the sending side datasets into. You might have to use zfs rename to move around receive-side datasets a little
> 
> If I understand this correctly, I can't directly sync the pool root 1:1 from production to backup? I have not been using a sub-tree for the replication.

You did understand this correctly, zrepl needs a sub-tree to receive into.
If you do want to start using zrepl, you might want to move the existing hierarchy into the new backup tree.

**Reply by gdevenyi** [reply]
Thanks @InsanePrawn 

So if I currently have a pool `data` with filesystems such as `data/quarantine` and a source of my backups `prod`, I should `zfs rename data/quarantine data/prod/quarantine`, assuming I set `root_fs: data`

One thing that's not quite clear to me yet, will zrepl properly take advantage of the non-zrepl snapshots which are already common in order to bootstrap the sync?

**Reply by InsanePrawn** [reply]
Yes, (unless you make a typo in the rename) you should either get an explicit replication conflict (no common ancestor snapshot) or an incremental receive.
(if you typo the rename, the dataset will not exist, and you'll get an initial full send.)

In any case, zrepl replication jobs will not just delete snapshots or datasets to get them out of the way.

One caveat: IIRC, the hierarchy will look like: `data/prod/name_of_prods_root_dataset/path/to/quarantine`, i.e. it'll try to mirror the source's hierarchy, inserting empty placeholder datasets where necessary

**Reply by gdevenyi** [reply]
> One caveat: IIRC, the hierarchy will look like: data/prod/name_of_prods_root_dataset/path/to/quarantine, i.e. it'll try to mirror the source's hierarchy, inserting empty placeholder datasets where necessary

Okay, so that means a filesystem on `prod` which is `data/quarantine` will show up as `data/prod/data/quarantine` for a sink with `root_fs: data` That makes sense as prod could have more than one pool.

**Reply by InsanePrawn** [reply]
Exactly :+1:

**Reply by problame** [reply]
@gdevenyi please report your experiences here so others who read this thread in the future are not left in doubt!

Also, we have started https://zrepl.github.io/usage.html#ops-runbooks , so, if you can find the time to write this up,  I'd be happy to add it to that page.
(ReStructured Text is preferred, but I'll also take your Markdown and convert it if necessary.)

**Reply by gdevenyi** [reply]
I've had one more problem. After performing the rename I now have a `zrepl/prod/data` filesystem which is empty, and has no common snapshot with the top-level filesystem of prod which is also being included in replication (but is empty), hence I get a STEP-ERROR.

Can I exclude the top-level somehow, or force an overwrite?

**Reply by gdevenyi** [reply]
This appears to exclude the top-level filesystem
```
  filesystems: {
    "<": true,
    "data": false
  }
```

**Reply by gdevenyi** [reply]
And now every replication is failing with:
```
cannot receive incremental stream: destination data/prod/data/${filesystem} has been modified
since most recent snapshot 
```

They haven't.

**Reply by gdevenyi** [reply]
A script to rollback all the datasets, which fixed it:
```sh
for filesystem in /data/prod/data/*; do
        zfs rollback "$(zfs list -t snapshot -s creation -o name data/prod/data/$(basename ${filesystem}) | tail -1)"
done
```

**Reply by gdevenyi** [reply]
@problame / @InsanePrawn 

I still had some problems with datasets being reported as modified, until I first unmounted the last few datasets and then rolled them back (despite nothing happening on the backup system...)

Are there any other properties the file systems should have set on them that my rename process outlined here wouldn't have handled and zrepl won't change? (I see something about the placeholder for example, and canmount or readyonly in some issues)

**Reply by problame** [reply]
@gdevenyi so, what I do is `canmount=off`, i.e., I don't mount the filesystems on the receiver at all.

**Comment by gdevenyi** [comment]
This feature would've removed the need for me to ask these questions, https://github.com/zrepl/zrepl/issues/94
